Csv 파싱 및 클래스 생성 자동화 기능 추가

GyutaeLee opened the pull request:
~~1. Csv 파일은 항상 "Resrouces/Csv/" 경로 하위에 배치해야 함
2. Csv Reader를 생성하면서 하위 Csv Path를 입력함
3. Read() 진행 : 따로 둔 이유는, 데이터가 필요하거나 읽고 싶은 시점에 읽기 위해서
4. GetInfoID & GetData 를 통해 데이터 파싱~~

**수정된 버전**
**1. "Resources/Csv" 경로에 Csv 파일 생성**

![image](https://user-images.githubusercontent.com/23232376/133458293-cf73c4be-9aed-4d32-8ce7-a083b5c82c45.png)

**2. Csv 양식**
- line 1 : 데이터 정보
- line 2 : 데이터 자료형
- line 3~ : 데이터
- 
![image](https://user-images.githubusercontent.com/23232376/133458809-cfac749d-278f-4bdb-92cc-0bfeec34b622.png)

**3. 상단 메뉴 [Csv -> Refresh All Csv Class from Csv Files] 클릭**

![image](https://user-images.githubusercontent.com/23232376/133458562-ddb1fd67-d7b4-4415-821b-bb8759f8c737.png)

**4. "Scripts/Util/Csv/Class" 경로에 Csv Class 가 생성되었는지 확인**

![image](https://user-images.githubusercontent.com/23232376/133458968-3d3428e5-97ed-4e0a-8499-d2730e24062b.png)

![image](https://user-images.githubusercontent.com/23232376/133459134-c205f9f3-8fa2-4eca-8486-f845ec32241e.png)

GyutaeLee committed:
feat : Csv Reader 클래스 추가
1. Csv 파일은 항상 "Resrouces/Csv/" 경로 하위에 배치해야 함
2. Csv Reader를 생성하면서 하위 Csv Path를 입력함
3. Read() 진행 : 따로 둔 이유는, 데이터가 필요하거나 읽고 싶은 시점에 읽기 위해서
4. GetInfoID & GetData 를 통해 데이터 파싱

Parkhyunseo commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+{
+    public class Reader
+    {
+        private static class Constants
```
내부에서만 사용하는 상수일 경우 따로 클래스 안빼기로 저번에 얘기했던 기억이 있습니다!

GyutaeLee replied:
기존 코드를 수정하면서 다른 곳에서 한 함수로 모은 걸 인지 못했네요!
수정했습니다.

Parkhyunseo commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+        public void Read()
+        {
+            string fullCsvPath = Constants.BaseCsvPath + this._csvPath;
+            TextAsset data = Resources.Load(fullCsvPath) as TextAsset;
```
TextAsset data = Resources.Load<TextAsset>(fullCsvPath);
형태를 선호하는데

둘 차이는 정확히 어떤지 모르겠네요

GyutaeLee replied:
Load에서 불러오는 형식이 UnityEngine.Object 인데 자동으로 변환이 되나보군요?!

그와 별개로 프로그래머가 무의식적으로 타입 변환을 놓쳐서 에러가 생기는 경우를 방지하기 위해 명시적으로 적어주는 것을 선호하는 편이긴 합니다.
c#에서는 대부분 자동으로 변환을 해주는걸까요?

이에 대해 회의 때 이야기를 나누어보면 좋을거 같습니다!

GyutaeLee replied:
TextAsset data = Resources.Load\<TextAsset>(fullCsvPath);

제네릭 형식으로 하는 것으로 합의했습니다. (컴파일 타임에 형변환 하도록)

Parkhyunseo commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+                return;
+            }
+
+            this._data = new List<Dictionary<string, string>>();
```
매번 생성하는 것보다 data를 생성해놓고 Clear하고 다시 담는 방식은 어떨까요

GyutaeLee replied:
관성적으로 참고 코드를 그대로 사용한 실수네요!
data 가 null인 경우에만 생성하고, null이 아닌 경우에는 Clear() 하도록 수정하겠습니다.

Parkhyunseo reviewed (commented):
이번에도 코멘트만 남겼습니다~

JuicyPark commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+            return int.Parse(this._data[index]["infoID"]);
+        }
+
+        public T GetData<T>(string key, int infoID)
```
제네릭 T에 대해 형식제한조건 where이 없어서 어떤클래스를 이용하든 사용이 가능할텐데 ex : GetData< Junsu > 이렇게 사용하고 아래에서 Convert 해줬을때 값이 정상적으로 보존이 되지 않을것 같습니다.

JuicyPark replied:
현재 이 클래스에서 data라는 이름의 자료형은 위에서도 말했듯 Dict<str,str> 인것 같은데 함수이름이 GetData로 하면 해당 자료형Dict<str,str> 반환 해야할것 같습니다.
1. data의 자료형을 바꾸거나
2. data 라는 이름을 바꾸거나
3. 해당 함수의 이름을 바꾸거나해야 혼동을 줄일수 있을것 같습니다.

GyutaeLee replied:
현재 csv reader에서는 int, long long, float, double, string 만을 읽어오려고 합니다.

where 형식을 사용하려고 구글링을 해보았는데,  해당 자료형에 관한 제약은 찾을 수가 없었습니다.
혹시 관련된 방법에 대해 조언을 해주실 수 있을까요?

JuicyPark commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        private string _csvPath;
+        private List<Dictionary<string, string>> _data;
```
아래 코드들을 봤을때 data의 자료형은 List<Dict<str.,str>>이 아닌 Dic<str,str> 인것 같아서 네이밍이 _datas가 맞는것 같습니다

GyutaeLee replied:
네! 그게 더 직관적인거 같습니다. 수정하겠습니다!

GyutaeLee committed:
fix : Constants 이너 클래스 제거
- 한 함수 내부에서 사용하기 때문에, 함수 내부로 선언 이동

GyutaeLee committed:
fix : Csv Reader 수정
- _data -> _datas 변수명 수정 : 복수의 데이터들이므로 명확하게 's' 추가
- _datas 생성 로직 변경 : 이미 생성이 되어 있으면 Clear() 하고 사용하도록 수정

GyutaeLee committed:
chore : 미완성된 코드 롤백

GyutaeLee committed:
feat : Csv 파싱 및 클래스 생성 자동화 기능 추가
- Csv/Reader.cs : Csv 데이터 파싱 클래스
- Csv/ClassMaker.cs : Csv 데이터 기반 클래스 자동 생성 클래스

JuicyPark commented on Union/Assets/Scripts/Util/Csv/ClassMaker.cs:
```diff
@@ -0,0 +1,131 @@
+            CheckAndCreateCsvClassDirectory(csvClassDirectoryPath);
+
+            string writePath = csvClassDirectoryPath + $"/{csvName}.cs";
+            File.WriteAllText(writePath, classStringBuilder.ToString());
```
![image](https://user-images.githubusercontent.com/31693348/133530076-9b08f5a5-919f-47e7-8a77-9a2c279a0d7f.png)
이쪽 라인에서 폴더못찾는 에러가 뜨는것 같습니다

GyutaeLee replied:
맥에서 작업할 때와 윈도우에서 경로를 불러오면 나오는 텍스트가 달라서 발생한 이슈였네요.
(윈도우에서는 '\(백슬래시)'를 불러오고, 맥에서는 '/'을 불러옴)

경로를 넣는 방식을 변경해서 해결했습니다.

JuicyPark commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,109 @@
+        public const int MinimumLineLengthOfCsv = 2;
+    }
+
+    public class Reader <T> where T : class
```
데이터를 참조하는 Monster가 100마리 있다고 하면, 100마리 Monster가 모두 new Reader< MonsterInfo >를 해서 데이터를 추출한뒤 딕셔너리에 넣는 로직을 거쳐야할것같습니다.

저는 이렇게 개선해보는걸 생각했는데 어떤까요? DataManager를 만들어서 1번 몬스터가 호출을하면 그때 딕셔너리에 저장을 해두면 2번 몬스터가 사용할때는 이미 DataManager에 데이터가 있기때문에 그걸 쓰면됨. (캐싱방법)
DataManager.Instance.GetData< MonsterInfo >( index ) 로 받아오거나
DataManager.Instance.GetDatas< MonsterInfo >( ) 해당 딕셔너리를 리스트로 전부 받아오는것도 있으면 좋을것 같습니다.

DataManager 속의 Dictionary는
Dictionary < string : 타입이름 , Dictionary < int , T > > 이런식으로 생각했습니다
(string은 생각이안나서 일단 넣은 예시이고 다른걸로도 구현가능할듯)

JuicyPark replied:
두번째로는
public class Reader <T> where T : class
가 어떤 class든 상관없이 접근이 가능한것 같아서 이런식으로 접근제한을 두는건 어떨까요?
![image](https://user-images.githubusercontent.com/31693348/133533548-6f80eddf-73af-4ec3-a801-675ec1cdc774.png)

GyutaeLee replied:
[1]번 코멘트
- 이번에 이 풀리퀘가 메인에 합쳐지면, Player와 Enemy에 csv 데이터를 붙이면서 따로 [Storage] 클래스를 만들어서 데이터를 저장(캐싱)한 후에 사용하는 코드를 추가할 예정이었습니다.
적어주신 방법과 비슷한 방식으로 진행하려고 설계 중이고, 디테일한 부분에 대해 고민하고 있는데 풀리퀘 후에, 다음 작업 때 추가하도록 하겠습니다.

[2]번 코멘트
- 아주 좋은 방법인거 같습니다! 바로 반영하겠습니다. 감사합니다.

JuicyPark commented on Union/Assets/Scripts/Util/Csv/Reader.cs:
```diff
@@ -0,0 +1,117 @@
+        public void Read()
+        {
+            string fullCsvPath = Constants.BaseCsvPath + this._csvPath;
+            TextAsset data = Resources.Load(fullCsvPath) as TextAsset;
```
이부분 저번 리뷰 반영되지 않은것 같습니다

GyutaeLee replied:
as 로 런타임에 불러오지 않고, 제네릭 형식으로 컴파일 타임에 불러오도록 수정했습니다.

GyutaeLee committed:
fix : csv 리소스 로딩 시점 변경
- 런타임 -> 컴파일 타임

GyutaeLee committed:
fix : csv reader 데이터 제한 변경
- IData 인터페이스를 상속하는 클래스만 사용할 수 있도록 수정

GyutaeLee committed:
fix : ClassMaker.cs 수정
- 폴더 경로 불러오는 로직 수정 : 운영 체제 별로 다를 수 있으므로, 경로를 텍스트로 고정시키고 csv 파일명만 추가로 입력함
- 기존에 있던 csv class들을 지우고, 새로 만들도록 기능 추가

GyutaeLee committed:
chore : 충돌 해결
